src/session_manager.rs

```diff
@@ -0,0 +1,1056 @@
+                )
+                .await
+            {
+                log::warn!("Failed to write file 'recently_used_sessions.json': {}", e);
+            }
+        };
+        glib::MainContext::default().spawn_local(future);
+    }
+
+    /// Within this function a new `Session` is created based on the passed client id. This session
+    /// is then added to the session stack.
+    pub fn add_logged_in_session(
+        &self,
+        client_id: i32,
+        database_dir_base_name: String,
+        on_test_dc: bool,
+        visible: bool,
+    ) {
+        let self_ = imp::SessionManager::from_instance(self);
+        let session = Session::new(self, client_id, database_dir_base_name, on_test_dc);
+
+        self_.sessions.add_child(&session);
+        session.set_sessions(&self_.sessions.pages());
+
+        if visible {
+            self_.sessions.set_visible_child(&session);
+            self_.main_stack.set_visible_child(&*self_.sessions);
+        }
+
+        self_
+            .clients
+            .borrow_mut()
+            .insert(client_id, ClientInfo::LoggedIn(session));
+
+        // Enable notifications for this client
+        RUNTIME.spawn(async move {
+            functions::SetOption::new()
+                .name("notification_group_count_max".to_string())
+                .value(enums::OptionValue::Integer(types::OptionValueInteger {
+                    value: 5,
+                }))
+                .send(client_id)
+                .await
+                .unwrap();
+        });
+    }
+
+    pub fn begin_chats_search(&self) {
+        if let Some(client_id) = self.active_logged_in_client_id() {
+            let self_ = imp::SessionManager::from_instance(self);
+            if let ClientInfo::LoggedIn(session) = self_.clients.borrow().get(&client_id).unwrap() {
+                session.begin_chats_search();
+            }
+        }
+    }
+}
+
+/// A struct for storing information about an sessions's database.
+#[derive(Debug)]
+pub struct DatabaseInfo {
+    // The base name of the database directory.
+    directory_base_name: String,
+    // Whether this database uses a test dc.
+    use_test_dc: bool,
+}
+
+/// A struct for representing the state of the data directory.
+#[derive(Debug)]
+pub enum DatadirState {
+    /// There are no sessions at all. This probably means that the application is started for the
+    /// first time or the data directory has been deleted by the user.
+    Empty,
+    /// There are several sessions logged in.
+    HasSessions {
+        /// The order of the recently used sessions that has been loaded from the data directory.
+        recently_used_sessions: Vec<String>,
+        database_infos: Vec<DatabaseInfo>,
+    },
+}
+
+/// Function for loading the recently used session from a file in the data directory. If an error
+/// occurs during reading, the empty set is returned.
+async fn load_recently_used_sessions() -> Vec<String> {
```

Comment:
Well, since `recently_used_sessions.json` just stores an array of strings why not using gsettings? We can create a new key of type "as" and using `gio::Settings::strv()` `gio::Settings::set_strv()` for managing it. It should also simplify things a bit.

Reply:
Yes, that makes sense. Initially, I had the idea to store the `recently used session` file per data directory, as the data directory could be changed by the `data-dir` cli parameter. 
But now the purpose of this parameter (quick testing things with other accounts) is invalid in a multi account capable application. 
So I will remove that parameter and will use gsettings instead of a file.